skfda/preprocessing/dim_reduction/_fpls.py

```diff
@@ -0,0 +1,829 @@
+            (The discretization weights matrix in the case of FDataGrid).
+
+        - G_cc: (n_targets, n_targets)
+            The inner product matrix for the Y block weights
+            (The discretization weights matrix in the case of FDataGrid).
+        - G_yc: (n_targets, n_targets)
+            The inner product matrix for the Y block data and weights
+            (The discretization weights matrix in the case of FDataGrid).
+
+        - L_X_inv: (n_features, n_features)
+            The inverse of the Cholesky decomposition:
+            L_X @ L_X.T = G_ww + P_x,
+            where P_x is a the penalty matrix for the X block.
+        - L_Y_inv: (n_targets, n_targets)
+            The inverse of the Cholesky decomposition:
+            L_Y @ L_Y.T = G_cc + P_y,
+            where P_y is a the penalty matrix for the Y block.
+
+        - deflation: The deflation method to use.
+            Can be "reg" for regression or "can" for dimensionality reduction.
+    Returns:
+        - W: (n_features, n_components)
+            The X block weights.
+        - C: (n_targets, n_components)
+            The Y block weights.
+        - T: (n_samples, n_components)
+            The X block scores.
+        - U: (n_samples, n_components)
+            The Y block scores.
+        - P: (n_features, n_components)
+            The X block loadings.
+        - Q: (n_targets, n_components)
+            The Y block loadings.
+    """
+    X = X.copy()
+    Y = Y.copy()
+    X = X - np.mean(X, axis=0)
+    Y = Y - np.mean(Y, axis=0)
+
+    if len(Y.shape) == 1:
+        Y = Y[:, np.newaxis]
+
+    # Store the matrices as list of columns
+    W, C = [], []
+    T, U = [], []
+    P, Q = [], []
+    for _ in range(n_components):
+        w, c = _calculate_weights(
+            X,
+            Y,
+            G_ww=G_ww,
+            G_xw=G_xw,
+            G_cc=G_cc,
+            G_yc=G_yc,
+            L_X_inv=L_X_inv,
+            L_Y_inv=L_Y_inv,
+        )
+
+        t = np.dot(X @ G_xw, w)
+        u = np.dot(Y @ G_yc, c)
+
+        p = np.dot(X.T, t) / (np.dot(t.T, t) + INV_EPS)
+
+        y_proyection = t if deflation == "reg" else u
+
+        q = np.dot(Y.T, y_proyection) / (
+            np.dot(y_proyection, y_proyection) + INV_EPS
+        )
+
+        X = X - np.outer(t, p)
+        Y = Y - np.outer(y_proyection, q)
+
+        W.append(w)
+        C.append(c)
+        T.append(t)
+        U.append(u)
+        P.append(p)
+        Q.append(q)
+
+    # Convert each least of columns to a matrix
+    return (  # noqa: WPS227 (too long output tuple)
+        np.array(W).T,
+        np.array(C).T,
+        np.array(T).T,
+        np.array(U).T,
+        np.array(P).T,
+        np.array(Q).T,
+    )
+
+
+BlockType = TypeVar(
+    "BlockType",
+    bound=Union[FDataGrid, FDataBasis, NDArrayFloat],
+)
+
+
+# Ignore too many public instance attributes
+class _FPLSBlock(Generic[BlockType]):  # noqa: WPS230
```

Comment:
I would rather have a Block abstract class and then separate classes implementing it, one for each kind of data. Then there will be no need to switch on the type of data inside each method.

Reply:
Done. Note that the factory function was still implemented with a switch. I tried to implement it using `multidispatch`. However, in that case, `mypy` does not identify the dependency between the type of `data` and the type of the returned `FPLSBlock`.

Reply:
It should, as `BlockType` is a `TypeVar`... What error did you get?

Reply:
I agree. It should. But it seems like mypy can no longer associate the BlockType variable to InputTypeX and InputTypeY, which leads to typing errors every time that a method from a block is called. Here are the errors:
```
skfda/preprocessing/dim_reduction/_fpls.py:814: error: Argument 1 to "transform" of "_FPLSBlock" has incompatible type "InputTypeX"; expected "BlockType"  [arg-type]
skfda/preprocessing/dim_reduction/_fpls.py:817: error: Argument 1 to "transform" of "_FPLSBlock" has incompatible type "InputTypeY"; expected "BlockType"  [arg-type]
skfda/preprocessing/dim_reduction/_fpls.py:839: error: Argument 1 to "transform" of "_FPLSBlock" has incompatible type "InputTypeX"; expected "BlockType"  [arg-type]
skfda/preprocessing/dim_reduction/_fpls.py:858: error: Argument 1 to "transform" of "_FPLSBlock" has incompatible type "InputTypeY"; expected "BlockType"  [arg-type]
skfda/preprocessing/dim_reduction/_fpls.py:886: error: Incompatible return value type (got "Tuple[BlockType, BlockType]", expected "Union[InputTypeX, Tuple[InputTypeX, InputTypeY]]")  [return-value]
skfda/preprocessing/dim_reduction/_fpls.py:888: error: Incompatible return value type (got "BlockType", expected "Union[InputTypeX, Tuple[InputTypeX, InputTypeY]]")  [return-value]
skfda/preprocessing/dim_reduction/_fpls.py:906: error: Incompatible return value type (got "BlockType", expected "InputTypeX")  [return-value]
skfda/preprocessing/dim_reduction/_fpls.py:924: error: Incompatible return value type (got "BlockType", expected "InputTypeY")  [return-value]
```
Note that the line numbers obviously do not match the contents of this pull request.
My implementation using `multimethod`is the following: 
```python
@multimethod.multidispatch
def _fpls_block_factory(
    data: BlockType,
    n_components: int,
    label: str,
    integration_weights: Union[NDArrayFloat, None],
    regularization: Union[L2Regularization[Any], None],
    weights_basis: Union[Basis, None],
) -> _FPLSBlock[BlockType]:
    raise TypeError("Invalid type for data")


@_fpls_block_factory.register
def _fpls_block_factory_array(
    data: np.ndarray,
    n_components: int,
    label: str,
    integration_weights: Union[np.ndarray, None],
    regularization: Union[L2Regularization[Any], None],
    weights_basis: Union[Basis, None],
) -> _FPLSBlock[NDArrayFloat]:
    return _FPLSBlockMultivariate(
        data=data,
        n_components=n_components,
        label=label,
    )


@_fpls_block_factory.register
def _fpls_block_factory_basis(
    data: FDataBasis,
    n_components: int,
    label: str,
    integration_weights: Union[np.ndarray, None],
    regularization: Union[L2Regularization[Any], None],
    weights_basis: Union[Basis, None],
) -> _FPLSBlock[FDataBasis]:
    return _FPLSBlockBasis(
        data=data,
        n_components=n_components,
        label=label,
        regularization=regularization,
        weights_basis=weights_basis,
    )


@_fpls_block_factory.register
def _fpls_block_factory_grid(
    data: FDataGrid,
    n_components: int,
    label: str,
    integration_weights: Union[np.ndarray, None],
    regularization: Union[L2Regularization[Any], None],
    weights_basis: Union[Basis, None],
) -> _FPLSBlock[FDataGrid]:
    return _FPLSBlockGrid(
        data=data,
        n_components=n_components,
        label=label,
        integration_weights=integration_weights,
        regularization=regularization,
    )

```